src/powerquery-formatter/serializeV2.ts

```diff
@@ -0,0 +1,566 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import * as PQP from "@microsoft/powerquery-parser";
+import { Ast } from "@microsoft/powerquery-parser/lib/powerquery-parser/language";
+
+import { SerializeParameterV2 } from "./themes";
+
+import { CommentCollection, CommentCollectionMap, LinearLengthMap, SerializeParameterMapV2 } from "./passes";
+import { IndentationLiteral, NewlineLiteral, TriedSerialize } from "./serialize";
+import { getLinearLengthV2 } from "./passes/utils/linearLengthV2";
+import { NodeKind } from "@microsoft/powerquery-parser/lib/powerquery-parser/language/ast/ast";
+
+// const GLOBAL_HEADING_WHITE_SPACE_REG: RegExp = /^[ \t]+/g;
+const GLOBAL_TAILING_WHITE_SPACE_REG: RegExp = /[ \t]+$/g;
+const GLOBAL_TAILING_CRLF_REG: RegExp = /(\r\n|\n\r|\r|\n)*$/g;
+
+export interface SerializeSettingsV2 extends PQP.CommonSettings {
+    readonly ast: Ast.TNode;
+    readonly nodeIdMapCollection: PQP.Parser.NodeIdMap.Collection;
+    readonly passthroughMaps: SerializePassthroughMapsV2;
+    readonly indentationLiteral: IndentationLiteral;
+    readonly newlineLiteral: NewlineLiteral;
+    readonly maxWidth?: number;
+}
+
+export interface SerializePassthroughMapsV2 {
+    readonly commentCollectionMap: CommentCollectionMap;
+    readonly containerIdHavingComments: Set<number>;
+    readonly serializeParameterMap: SerializeParameterMapV2;
+}
+
+export function trySerializeV2(settings: SerializeSettingsV2): Promise<TriedSerialize> {
+    return PQP.ResultUtils.ensureResultAsync(() => serializeV2(settings), settings.locale);
+}
+
+type LastTokenType = "Opener" | "Closer" | "Divider" | "SpanContent" | "Null" | "CommentsLine";
```

Comment:
Can we instead use a const enum? It should compile down to the same so there's not much benefit, but it's more consistent with the rest of the codebase.

Reply:
sure